Argo/Globals/StandardTypes.swift

```diff
@@ -2,53 +2,60 @@ import Foundation
 import Runes
 
 extension String: JSONDecodable {
-  public static func decode(j: JSON) -> String? {
+  public static func fromJSON(j: JSON) -> Parser<String> {
     switch j {
-    case let .String(s): return s
-    default: return .None
+    case let .String(s): return pure(s)
+    default: return .typeMismatch("String", object: j)
     }
   }
 }
 
 extension Int: JSONDecodable {
-  public static func decode(j: JSON) -> Int? {
+  public static func fromJSON(j: JSON) -> Parser<Int> {
     switch j {
-    case let .Number(n): return n as Int
-    default: return .None
+    case let .Number(n): return pure(n as Int)
+    default: return .typeMismatch("Int", object: j)
     }
   }
 }
 
 extension Double: JSONDecodable {
-  public static func decode(j: JSON) -> Double? {
+  public static func fromJSON(j: JSON) -> Parser<Double> {
     switch j {
-    case let .Number(n): return n as Double
-    default: return .None
+    case let .Number(n): return pure(n as Double)
+    default: return .typeMismatch("Double", object: j)
     }
   }
 }
 
 extension Bool: JSONDecodable {
-  public static func decode(j: JSON) -> Bool? {
+  public static func fromJSON(j: JSON) -> Parser<Bool> {
     switch j {
-    case let .Number(n): return n as Bool
-    default: return .None
+    case let .Number(n): return pure(n as Bool)
+    default: return .typeMismatch("Bool", object: j)
     }
   }
 }
 
 extension Float: JSONDecodable {
-  public static func decode(j: JSON) -> Float? {
+  public static func fromJSON(j: JSON) -> Parser<Float> {
     switch j {
-    case let .Number(n): return n as Float
-    default: return .None
+    case let .Number(n): return pure(n as Float)
+    default: return .typeMismatch("Float", object: j)
     }
   }
 }
 
-public func decodeArray<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> [A]? {
+public func decodeArray<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> Parser<[A]> {
   switch value {
-  case let .Array(a): return sequence({ A.decode($0) } <^> a)
-  default: return .None
+  case let .Array(a): return sequence({ A.fromJSON($0) } <^> a)
+  default: return .typeMismatch("Array", object: value)
+  }
+}
+
+public func decodeObject<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> Parser<[String: A]> {
+  switch value {
+  case let .Object(o): return sequence(map(o, A.fromJSON))
```

Comment:
the partial application of `A.fromJSON` might cause segfaults on 32 bit systems right now. Do we care? We aren't cutting a new release till 1.2 is out, at which point that bug should be fixed anyway.